File: services/graphql-wms/src/entities/locations.rs
use crate::entities::_generated::locations;
use crate::entities::_generated::sea_orm_active_enums::LocationTypeEnum;
use async_graphql::InputObject;
use sea_orm::{
    ActiveModelBehavior,
    ActiveValue::{NotSet, Set},
    IntoActiveModel,
};
use uuid::Uuid;

#[derive(Debug, Clone, InputObject)]
pub struct InsertLocation {
    pub warehouse_id: Uuid,
    pub parent_location_id: Option<Uuid>,
    pub name: String,
    pub barcode: Option<String>,
    pub r#type: LocationTypeEnum,
    pub level: Option<i32>,
    pub path: Option<String>,
    pub max_weight: Option<f32>,
    pub max_volume: Option<f32>,
    pub max_pallets: Option<i32>,
    pub x_coordinate: Option<f32>,
    pub y_coordinate: Option<f32>,
    pub z_coordinate: Option<f32>,
    pub is_pickable: Option<bool>,
    pub is_receivable: Option<bool>,
    pub temperature_controlled: Option<bool>,
    pub hazmat_approved: Option<bool>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Clone, InputObject)]
pub struct UpdateLocation {
    pub warehouse_id: Option<Uuid>,
    pub parent_location_id: Option<Option<Uuid>>,
    pub name: Option<String>,
    pub barcode: Option<Option<String>>,
    pub r#type: Option<LocationTypeEnum>,
    pub level: Option<Option<i32>>,
    pub path: Option<Option<String>>,
    pub max_weight: Option<Option<f32>>,
    pub max_volume: Option<Option<f32>>,
    pub max_pallets: Option<Option<i32>>,
    pub x_coordinate: Option<Option<f32>>,
    pub y_coordinate: Option<Option<f32>>,
    pub z_coordinate: Option<Option<f32>>,
    pub is_pickable: Option<Option<bool>>,
    pub is_receivable: Option<Option<bool>>,
    pub temperature_controlled: Option<Option<bool>>,
    pub hazmat_approved: Option<Option<bool>>,
    pub is_active: Option<Option<bool>>,
}

impl IntoActiveModel<locations::ActiveModel> for InsertLocation {
    fn into_active_model(self) -> locations::ActiveModel {
        let mut active_model = locations::ActiveModel::new();
        active_model.warehouse_id = Set(self.warehouse_id);
        active_model.parent_location_id = Set(self.parent_location_id);
        active_model.name = Set(self.name);
        active_model.barcode = Set(self.barcode);
        active_model.r#type = Set(self.r#type);
        active_model.level = Set(self.level);
        active_model.path = Set(self.path);
        active_model.max_weight = Set(self.max_weight);
        active_model.max_volume = Set(self.max_volume);
        active_model.max_pallets = Set(self.max_pallets);
        active_model.x_coordinate = Set(self.x_coordinate);
        active_model.y_coordinate = Set(self.y_coordinate);
        active_model.z_coordinate = Set(self.z_coordinate);
        active_model.is_pickable = Set(self.is_pickable);
        active_model.is_receivable = Set(self.is_receivable);
        active_model.temperature_controlled = Set(self.temperature_controlled);
        active_model.hazmat_approved = Set(self.hazmat_approved);
        active_model.is_active = Set(self.is_active);
        active_model
    }
}

impl IntoActiveModel<locations::ActiveModel> for UpdateLocation {
    fn into_active_model(self) -> locations::ActiveModel {
        let mut active_model = locations::ActiveModel::new();
        active_model.warehouse_id = self.warehouse_id.map(Set).unwrap_or(NotSet);
        active_model.parent_location_id = self.parent_location_id.map(Set).unwrap_or(NotSet);
        active_model.name = self.name.map(Set).unwrap_or(NotSet);
        active_model.barcode = self.barcode.map(Set).unwrap_or(NotSet);
        active_model.r#type = self.r#type.map(Set).unwrap_or(NotSet);
        active_model.level = self.level.map(Set).unwrap_or(NotSet);
        active_model.path = self.path.map(Set).unwrap_or(NotSet);
        active_model.max_weight = self.max_weight.map(Set).unwrap_or(NotSet);
        active_model.max_volume = self.max_volume.map(Set).unwrap_or(NotSet);
        active_model.max_pallets = self.max_pallets.map(Set).unwrap_or(NotSet);
        active_model.x_coordinate = self.x_coordinate.map(Set).unwrap_or(NotSet);
        active_model.y_coordinate = self.y_coordinate.map(Set).unwrap_or(NotSet);
        active_model.z_coordinate = self.z_coordinate.map(Set).unwrap_or(NotSet);
        active_model.is_pickable = self.is_pickable.map(Set).unwrap_or(NotSet);
        active_model.is_receivable = self.is_receivable.map(Set).unwrap_or(NotSet);
        active_model.temperature_controlled =
            self.temperature_controlled.map(Set).unwrap_or(NotSet);
        active_model.hazmat_approved = self.hazmat_approved.map(Set).unwrap_or(NotSet);
        active_model.is_active = self.is_active.map(Set).unwrap_or(NotSet);
        active_model
    }
}

File: src/repositories/wms/locations.ts
import {
  DeleteQueryBuilder,
  DeleteResult,
  Insertable,
  InsertQueryBuilder,
  Kysely,
  OrderByExpression,
  OrderByModifiers,
  SelectExpression,
  SelectQueryBuilder,
  Updateable,
  UpdateQueryBuilder,
} from 'kysely';
import { DB, WmsLocationTypeEnum } from '@/db/types';
import { FilterConfig, GenericRepository, SortConfig } from '../interface';

export class LocationRepository implements GenericRepository<'wms.locations'> {
  constructor(private db: Kysely<DB>) {}

  paginate(
    page?: number,
    limit?: number,
    sort?: SortConfig<'wms.locations'> | undefined,
    filter?: FilterConfig<'wms.locations'> | undefined,
  ): SelectQueryBuilder<
    DB,
    'wms.locations',
    {
      barcode: string | null;
      createdAt: Date | null;
      hazmatApproved: boolean | null;
      id: string;
      isActive: boolean | null;
      isPickable: boolean | null;
      isReceivable: boolean | null;
      level: number | null;
      maxPallets: number | null;
      maxVolume: number | null;
      maxWeight: number | null;
      name: string;
      parentLocationId: string | null;
      path: string | null;
      temperatureControlled: boolean | null;
      type: WmsLocationTypeEnum;
      updatedAt: Date | null;
      warehouseId: string;
      xCoordinate: number | null;
      yCoordinate: number | null;
      zCoordinate: number | null;
    }
  > {
    let query = this.db.selectFrom('wms.locations').selectAll();

    if (limit) query = query.limit(limit);

    if (page && limit) query = query.offset((page - 1) * limit);

    for (const sortCol of sort || []) {
      query = query.orderBy(sortCol.column, sortCol.order);
    }

    for (const filterCol of filter || []) {
      query = query.where(
        filterCol.column,
        filterCol.operation,
        filterCol.value,
      );
    }

    return query;
  }
  range(
    from: Date,
    to: Date,
    sort?: SortConfig<'wms.locations'> | undefined,
    filter?: FilterConfig<'wms.locations'> | undefined,
  ): SelectQueryBuilder<
    DB,
    'wms.locations',
    {
      barcode: string | null;
      createdAt: Date | null;
      hazmatApproved: boolean | null;
      id: string;
      isActive: boolean | null;
      isPickable: boolean | null;
      isReceivable: boolean | null;
      level: number | null;
      maxPallets: number | null;
      maxVolume: number | null;
      maxWeight: number | null;
      name: string;
      parentLocationId: string | null;
      path: string | null;
      temperatureControlled: boolean | null;
      type: WmsLocationTypeEnum;
      updatedAt: Date | null;
      warehouseId: string;
      xCoordinate: number | null;
      yCoordinate: number | null;
      zCoordinate: number | null;
    }
  > {
    let query = this.db
      .selectFrom('wms.locations')
      .selectAll()
      .where('createdAt', '>=', from)
      .where('createdAt', '<=', to);

    for (const sortCol of sort || []) {
      query = query.orderBy(sortCol.column, sortCol.order);
    }

    for (const filterCol of filter || []) {
      query = query.where(
        filterCol.column,
        filterCol.operation,
        filterCol.value,
      );
    }

    return query;
  }
  in(
    values: string[],
  ): SelectQueryBuilder<
    DB,
    'wms.locations',
    {
      barcode: string | null;
      createdAt: Date | null;
      hazmatApproved: boolean | null;
      id: string;
      isActive: boolean | null;
      isPickable: boolean | null;
      isReceivable: boolean | null;
      level: number | null;
      maxPallets: number | null;
      maxVolume: number | null;
      maxWeight: number | null;
      name: string;
      parentLocationId: string | null;
      path: string | null;
      temperatureControlled: boolean | null;
      type: WmsLocationTypeEnum;
      updatedAt: Date | null;
      warehouseId: string;
      xCoordinate: number | null;
      yCoordinate: number | null;
      zCoordinate: number | null;
    }
  > {
    return this.db
      .selectFrom('wms.locations')
      .selectAll()
      .where('id', 'in', values);
  }
  create(
    value: { name: string; type: WmsLocationTypeEnum; warehouseId: string } & {
      barcode?: string | null | undefined;
      createdAt?: string | Date | null | undefined;
      hazmatApproved?: boolean | null | undefined;
      id?: string | undefined;
      isActive?: boolean | null | undefined;
      isPickable?: boolean | null | undefined;
      isReceivable?: boolean | null | undefined;
      level?: number | null | undefined;
      maxPallets?: number | null | undefined;
      maxVolume?: number | null | undefined;
      maxWeight?: number | null | undefined;
      parentLocationId?: string | null | undefined;
      path?: string | null | undefined;
      temperatureControlled?: boolean | null | undefined;
      updatedAt?: string | Date | null | undefined;
      xCoordinate?: number | null | undefined;
      yCoordinate?: number | null | undefined;
      zCoordinate?: number | null | undefined;
    },
  ): InsertQueryBuilder<
    DB,
    'wms.locations',
    {
      barcode: string | null;
      createdAt: Date | null;
      hazmatApproved: boolean | null;
      id: string;
      isActive: boolean | null;
      isPickable: boolean | null;
      isReceivable: boolean | null;
      level: number | null;
      maxPallets: number | null;
      maxVolume: number | null;
      maxWeight: number | null;
      name: string;
      parentLocationId: string | null;
      path: string | null;
      temperatureControlled: boolean | null;
      type: WmsLocationTypeEnum;
      updatedAt: Date | null;
      warehouseId: string;
      xCoordinate: number | null;
      yCoordinate: number | null;
      zCoordinate: number | null;
    }
  > {
    return this.db.insertInto('wms.locations').values(value).returningAll();
  }
  update(
    id: string,
    value: {
      barcode?: string | null | undefined;
      createdAt?: string | Date | null | undefined;
      hazmatApproved?: boolean | null | undefined;
      id?: string | undefined;
      isActive?: boolean | null | undefined;
      isPickable?: boolean | null | undefined;
      isReceivable?: boolean | null | undefined;
      level?: number | null | undefined;
      maxPallets?: number | null | undefined;
      maxVolume?: number | null | undefined;
      maxWeight?: number | null | undefined;
      name?: string | undefined;
      parentLocationId?: string | null | undefined;
      path?: string | null | undefined;
      temperatureControlled?: boolean | null | undefined;
      type?: WmsLocationTypeEnum | undefined;
      updatedAt?: string | Date | null | undefined;
      warehouseId?: string | undefined;
      xCoordinate?: number | null | undefined;
      yCoordinate?: number | null | undefined;
      zCoordinate?: number | null | undefined;
    },
  ): UpdateQueryBuilder<
    DB,
    'wms.locations',
    'wms.locations',
    {
      barcode: string | null;
      createdAt: Date | null;
      hazmatApproved: boolean | null;
      id: string;
      isActive: boolean | null;
      isPickable: boolean | null;
      isReceivable: boolean | null;
      level: number | null;
      maxPallets: number | null;
      maxVolume: number | null;
      maxWeight: number | null;
      name: string;
      parentLocationId: string | null;
      path: string | null;
      temperatureControlled: boolean | null;
      type: WmsLocationTypeEnum;
      updatedAt: Date | null;
      warehouseId: string;
      xCoordinate: number | null;
      yCoordinate: number | null;
      zCoordinate: number | null;
    }
  > {
    return this.db
      .updateTable('wms.locations')
      .set(value)
      .where('id', '=', id)
      .returningAll();
  }
  delete(id: string): DeleteQueryBuilder<DB, 'wms.locations', DeleteResult> {
    return this.db.deleteFrom('wms.locations').where('id', '=', id);
  }
}

export class WmsLocationRepository {
  constructor(private db: Kysely<DB>) {}

  select(
    page: number,
    perPage: number,
    fields?: SelectExpression<DB, 'wms.locations'>,
    search?: string,
    sort?: {
      field: OrderByExpression<DB, 'wms.locations', {}>;
      order: OrderByModifiers;
    }[],
  ) {
    let builder = this.db
      .selectFrom('wms.locations')
      .limit(perPage)
      .offset((page - 1) * perPage);

    if (fields) {
      builder = builder.select(fields);
    } else {
      builder = builder.selectAll();
    }

    // sort
    for (const field of sort || []) {
      builder = builder.orderBy(field.field, field.order);
    }

    if (search)
      builder = builder.where('wms.locations.id', 'like', `%${search}%`);

    return builder;
  }

  create(value: Insertable<DB['wms.locations']>) {
    return this.db.insertInto('wms.locations').values(value).returningAll();
  }

  batchCreate(values: Insertable<DB['wms.locations']>[]) {
    return this.db.insertInto('wms.locations').values(values).returningAll();
  }

  update(
    id: DB['wms.locations']['id']['__update__'],
    value: Updateable<DB['wms.locations']>,
  ) {
    return this.db
      .updateTable('wms.locations')
      .set(value)
      .where('wms.locations.id', '=', id)
      .returningAll();
  }

  delete(id: DB['wms.locations']['id']['__update__']) {
    return this.db
      .deleteFrom('wms.locations')
      .where('wms.locations.id', '=', id);
  }
}
